### PR TITLE
Update use_role testing in test_run_processor

### PR DESCRIPTION
**Pull request by sfc-gh-pjafari**
### Pre-review checklist
   * [ ] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [ ] I've added or updated automated unit tests to verify correctness of my new code.
   * [ ] I've added or updated integration tests to verify correctness of my new code.
   * [ ] I've confirmed that my changes are working by executing CLI's commands manually on MacOS.
   * [ ] I've confirmed that my changes are working by executing CLI's commands manually on Windows.
   * [ ] I've confirmed that my changes are up-to-date with the target branch.
   * [ ] I've described my changes in the release notes.
   * [ ] I've described my changes in the section below.

### Changes description
This is the first (small) step toward improvements to unit tests. [Epic link](https://snowflakecomputing.atlassian.net/browse/SNOW-1528911)

A WIP document of areas of improvement I've found so far [[WIP] Current issues found in SnowCli tests](https://docs.google.com/document/d/1rPoKoQO3BYPUBRrRBY2dYtM-bWdh38ZCejOwrq1GTmY/edit?usp=sharing)

Example to show how we can update testing of use_role, so we can write less boilerplate. Using run_processor testing as an example.
- Add tests for code paths that have use_role (2 in this case)
- Mock use_role in all other tests
...

### CODE REVIEW

File: tests/nativeapp/test_run_processor.py [hunk 1]
```diff
@@ -90,12 +91,11 @@ def _get_na_run_processor():
...
-
-# Test create_dev_app with exception thrown trying to use the warehouse
+# Test create_dev_app with no existing application AND create succeeds AND app role == package role
+@mock.patch(RUN_PROCESSOR_GET_EXISTING_APP_INFO, return_value=None)
 @mock.patch(NATIVEAPP_MANAGER_EXECUTE)
-@mock_connection()
-def test_create_dev_app_w_warehouse_access_exception(
-    mock_conn, mock_execute, temp_dir, mock_cursor
+def test_use_role_in_create_dev_app_w_no_additional_privileges(
+    mock_execute, mock_get_existing_app_info, temp_dir, mock_cursor
```
Comment (sfc-gh-pjafari) [inline tests/nativeapp/test_run_processor.py:98]: this `test_use_role_in_create_dev_app_w_no_additional_privileges` and the below `test_use_role_in_create_dev_app_w_additional_privileges ` are the two that test use_role flows in `create_or_upgrade`.

File: tests/nativeapp/test_run_processor.py [hunk 2]
```diff
@@ -184,29 +198,78 @@ def test_create_dev_app_create_new_w_no_additional_privileges(
...
+    assert mock_execute_query.mock_calls == mock_execute_query_expected
+
+
+# Test create_dev_app with exception thrown trying to use the warehouse
+@mock.patch(NATIVEAPP_MANAGER_EXECUTE)
+@mock.patch(NATIVEAPP_MANAGER_USE_ROLE)
+@mock_connection()
+def test_create_dev_app_w_warehouse_access_exception(
+    mock_conn, mock_use_role, mock_execute, temp_dir, mock_cursor
+):
```
Comment (sfc-gh-pjafari) [inline tests/nativeapp/test_run_processor.py:223]: In this test and the one below, we mock use_role with`@mock.patch(NATIVEAPP_MANAGER_USE_ROLE)`. The mocking of `_execute_query` that was made as part of use_role are removed from `mock_execute_helper`.

File: tests/nativeapp/test_manager.py [hunk 3]
```diff
@@ -247,26 +249,21 @@ def test_get_app_pkg_distribution_in_snowflake(mock_execute, temp_dir, mock_curs
...
+    mock_execute, mock_use_role, temp_dir, mock_cursor
 ):
 
     side_effects, expected = mock_execute_helper(
         [
-            (
-                mock_cursor([{"CURRENT_ROLE()": "old_role"}], []),
-                mock.call("select current_role()", cursor_class=DictCursor),
-            ),
-            (None, mock.call("use role package_role")),
```
Comment (sfc-gh-bdufour) [inline tests/nativeapp/test_manager.py:261]: hmm, but not executing "use role" at the right time means that the behaviour could be incorrect. Removing boilerplate is good, but aren't we making the test weaker as a result?

Comment (sfc-gh-pjafari) [reply tests/nativeapp/test_manager.py:261]: the idea is we only need to test the behavior of `use_role` once for each occurrence of it. So if we are writing tests for one function (that uses `use_role`) we can write one (or as many needed) test to make sure that `use_role` works as expected. The rest of the tests we write for this function can mock `use_role`.

Comment (sfc-gh-pjafari) [reply tests/nativeapp/test_manager.py:261]: In this example, `use_role` behavior is tested in `test_get_app_pkg_distribution_in_snowflake`. The rest of the tests that are written for `get_app_pkg_distribution_in_snowflake` in particular, can mock `use_role`.

Comment (sfc-gh-bdufour) [reply tests/nativeapp/test_manager.py:261]: For a happy path vs exception path pair of tests, that might make sense in some cases, but I'm not sure I'm generally aligned with this idea. And even then, it seems like a micro-optimization to me. As Cam was saying earlier today, DRY isn't necessary a target in tests, and often will end up with negative side effects. Happy to discuss more when I'm back if you want.

**Comment by sfc-gh-pjafari** [comment]
Closing this in favor of longer-term solution of refactoring the implementation.

**Pull request closed without merging**